### PR TITLE
Add "Auto" value in option parsing

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Options/Auto.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+#include <optional>
+#include <ostream>
+#include <string>
+
+#include "Options/Options.hpp"
+
+namespace Options {
+/// \ingroup OptionParsingGroup
+/// A class indicating that a parsed value can be automatically
```
Comment (nilsvu) [inline src/Options/Auto.hpp]: This paragraph should probably be `\brief`

File: src/Options/Auto.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+/// specified either as "Auto" or as a value of type `T`.  When this
+/// class is passed to the constructor of the class taking it as an
+/// option, it can be implicitly converted to a `std::optional<T>`.
+///
+/// \snippet Test_Auto.cpp example_class
+/// \snippet Test_Auto.cpp example_create
+template <typename T>
+class Auto {
+ public:
+  Auto() = default;
```
Comment (nilsvu) [inline src/Options/Auto.hpp:29]: noexcept?

File: src/Options/Auto.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <optional>
+#include <ostream>
+#include <string>
```
Comment (nilsvu) [inline src/Options/Auto.hpp:8]: include `utility`

File: src/Options/Auto.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  operator const std::optional<T>&() const { return value_; }
+
+ private:
+  std::optional<T> value_{};
+};
+
+template <typename T>
+std::ostream& operator<<(std::ostream& s, const Auto<T>& x) noexcept {
```
Comment (nilsvu) [inline src/Options/Auto.hpp]: (nitpick) in other classes we call `s` -> `os`

File: tests/Unit/Options/Test_Auto.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    CHECK(created);
+    CHECK(*created);
+  }
+
+  check_create<std::vector<int>>("Auto", {});
+  check_create<std::vector<int>>("", std::vector<int>{});
+  check_create<std::vector<int>>(
+      "- 1\n"
+      "- 2\n"
+      "- 3\n",
```
Comment (nilsvu) [inline tests/Unit/Options/Test_Auto.cpp]: you could also write this as `"[1, 2, 3]"`

**Review by nilsvu — CHANGES_REQUESTED** [review]
This looks great! A few minor comments:

File: src/Options/Auto.hpp (hunk 2 above)
Comment (wthrowe) [reply src/Options/Auto.hpp:29]: The meaning of writing noexcept there varies between versions of the standard, and clang implements it incorrectly for C++17, so I'm avoiding it these days.  Leaving it out always does the right thing.  (Same with constexpr.)

**Comment by nilsvu** [comment]
LGTM please squash

**Comment by wthrowe** [comment]
Squashed.  I also updated the second commit to also do the CCE end time, which I missed before.  The changes were virtually identical to those for the start time (and not very long), so I squashed them in.

**Review by nilsvu — DISMISSED** [review]

**Review by moxcodes — DISMISSED** [review]
Looks great, thanks for adding this!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
The build fails (I think partly because the compilers are being silly), and there is an actual clang-tidy warning

**Comment by wthrowe** [comment]
The fixes for CI were at least slightly nontrivial, so I pushed them as a fixup.

**Comment by nilsdeppe** [comment]
Looks good. I'm happy for you to squash :)

**Review by moxcodes — DISMISSED** [review]
I'm happy for you to squash.

File: tests/Unit/Options/Test_Auto.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+ public:
+  ExampleClass() = default;
+
+  struct AutoArg {
+    using type = Options::Auto<int>;
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ >= 8 && __GNUC__ < 11
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif  // defined(__GNUC__) && !defined(__clang__) && __GNUC__ => 8 && __GNUC__ < 11
+    static type default_value() noexcept { return {}; }
```
Comment (nilsvu) [inline tests/Unit/Options/Test_Auto.cpp:103]: - Do you have an idea why GCC doesn't like this instantiation, but is fine with the one in the CCE code? It's a bit unfortunate that the pragmas get printed to the documentation snippet.
- You add pragmas for GCC 8--9 in Auto.hpp, and for GCC 8--10 here. Is that on purpose?

Please squash directly if you end up making any changes.

Comment (wthrowe) [reply tests/Unit/Options/Test_Auto.cpp:103]: I have no idea why different version of GCC complain about different variations on that line.

Good point about the documentation.  I've moved the pragmas outside the snippet.

**Review by nilsdeppe — APPROVED** [review]

File: tests/Unit/Options/Test_Auto.cpp (hunk 6 above)
Comment (nilsvu) [reply tests/Unit/Options/Test_Auto.cpp:103]: So GCC 10 indeed complains about the one line but not the other, whereas GCC 8 & 9 complain about both? that's weird

**Review by nilsvu — APPROVED** [review]

**Pull request merged**